src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs

```diff
@@ -278,26 +286,90 @@ public async Task StartAsync(Uri url, TransferFormat transferFormat, Cancellatio
 
         Log.StartedTransport(_logger);
 
-        // Create the pipe pair (Application's writer is connected to Transport's reader, and vice versa)
-        var pair = DuplexPipe.CreateConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+        _stopCts = new CancellationTokenSource();
 
-        _transport = pair.Transport;
-        _application = pair.Application;
+        var ignoreFirstCanceled = false;
+
+        if (_transport is null)
+        {
+            // Create the pipe pair (Application's writer is connected to Transport's reader, and vice versa)
+            DuplexPipePair pair;
+            if (_useAck)
+            {
+                pair = CreateAckConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+            }
+            else
+            {
+                pair = DuplexPipe.CreateConnectionPair(_httpConnectionOptions.TransportPipeOptions, _httpConnectionOptions.AppPipeOptions);
+            }
+
+            _transport = pair.Transport;
+            _application = pair.Application;
+        }
+        else
+        {
+            if (_application!.Input is AckPipeReader reader)
+            {
+                if (reader.Resend())
+                {
+                    // Start reconnect ack handshake
+                    // 1. Send ack ID to server for last message we recieved from server before we disconnected
+                    // 2. Read from server to get the last ack ID it received before we disconnecting
+                    // 3. Resume normal send/receive loops
+
+                    ignoreFirstCanceled = true;
+                    var buf = new byte[AckPipeWriter.FrameSize];
```

Comment:
naming clarity: is this `FrameHeaderSize`? `ReconnectFrameSize`? something else?

Reply:
It is only on reconnect which should be rare, unless you have a terrible internet connection.